In [2]:
import numpy as np
import pandas as pd

# Problem 4: Document Clustering

Document clustering is an important application of clustering algorithms.
Document clustering algorithms can be used to automatically group search results into categories, making it easier to find relevant results.

In this problem, you'll be clustering articles published on [Daily Kos](https://www.dailykos.com/), an American political blog

The file *dailykos.csv* contains data on 3,430 news articles or blogs that have been posted on Daily Kos.
These articles were posted in 2004, leading up to the United States Presidential Election. 

In [4]:
# load the data
url = 'https://raw.githubusercontent.com/um-perez-alvaro/Data-Science-Theory/master/Data/dailykos.csv'
data = pd.read_csv(url)
data

abandon  abc  ability  abortion  absolute  abstain  abu  abuse  accept  \
0           0    0        0         0         0        0    0      0       0   
1           0    0        0         0         0        0    0      0       0   
2           0    0        0         0         0        1    0      0       0   
3           0    0        0         0         0        0    0      0       0   
4           0    0        0         0         0        0    0      0       0   
...       ...  ...      ...       ...       ...      ...  ...    ...     ...   
3425        0    0        0         0         0        0    0      0       0   
3426        0    0        0         0         0        0    0      0       0   
3427        0    0        0         0         0        0    0      0       0   
3428        0    0        0         0         0        0    0      0       0   
3429        0    0        0         0         0        0    0      0       0   

      access  ...  yeah  year  yesterday  york  youll  young  youre  youve  \
0          0  ...     0     0          0     0      0      0      0      0   
1          0  ...     0     0          0     0      0      0      0      0   
2          0  ...     0     0          0     0      0      0      0      0   
3          0  ...     0     0          2     0      0      1      0      0   
4          0  ...     0     1          1     0      0      1      0      0   
...      ...  ...   ...   ...        ...   ...    ...    ...    ...    ...   
3425       0  ...     0     0          0     0      0      0      0      0   
3426       0  ...     0     0          0     0      0      0      0      0   
3427       0  ...     0     0          0     1      0      0      0      0   
3428       0  ...     0     0          0     0      0      0      0      0   
3429       0  ...     0     0          0     0      0      0      0      0   

      zogby  zone  
0         0     1  
1         0     0  
2         0     0  
3         0     0  
4         0     0  
...     ...   ...  
3425      0     0  
3426      0     0  
3427      0     0  
3428      0     0  
3429      0     0  

[3430 rows x 1545 columns]

Each of the features in the dataset is a word that has appeared in at least 50 different articles (1,545 words in total).  For each document, the feature values are the number of times that word appeared in the document.